source/secure_tunneling_operations.c

```diff
@@ -0,0 +1,670 @@
+    if (aws_byte_buf_init(&message_storage->storage, allocator, storage_capacity)) {
+        return AWS_OP_ERR;
+    }
+
+    struct aws_secure_tunnel_message_view *storage_view = &message_storage->storage_view;
+
+    storage_view->type = message_options->type;
+    storage_view->ignorable = message_options->ignorable;
+    storage_view->stream_id = message_options->stream_id;
+
+    switch (type) {
+        case AWS_STOT_DATA:
+            storage_view->type = AWS_SECURE_TUNNEL_MT_DATA;
+            break;
+        case AWS_STOT_STREAM_START:
+            storage_view->type = AWS_SECURE_TUNNEL_MT_STREAM_START;
+            break;
+        case AWS_STOT_STREAM_RESET:
+            storage_view->type = AWS_SECURE_TUNNEL_MT_STREAM_RESET;
+            break;
+        default:
+            storage_view->type = AWS_SECURE_TUNNEL_MT_UNKNOWN;
+            break;
+    }
+
+    storage_view->service_id = message_options->service_id;
+    if (aws_byte_buf_append_and_update(&message_storage->storage, &storage_view->service_id)) {
+        return AWS_OP_ERR;
+    }
+
+    storage_view->payload = message_options->payload;
+    if (aws_byte_buf_append_and_update(&message_storage->storage, &storage_view->payload)) {
+        return AWS_OP_ERR;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+void aws_secure_tunnel_message_storage_clean_up(struct aws_secure_tunnel_message_storage *message_storage) {
+    aws_byte_buf_clean_up(&message_storage->storage);
+}
+
+/* Sets the stream id on outbound message based on the service id (or lack of for V1) to the current one being used. */
+static int s_aws_secure_tunnel_operation_message_set_stream_id(
+    struct aws_secure_tunnel_operation *operation,
+    struct aws_secure_tunnel *secure_tunnel) {
+
+    struct aws_secure_tunnel_operation_message *message_op = operation->impl;
+    int32_t stream_id = INVALID_STREAM_ID;
+
+    struct aws_secure_tunnel_message_view *message_view = &message_op->options_storage.storage_view;
+
+    if (message_view->service_id.len > 0) {
+        struct aws_string *service_id = NULL;
+        service_id = aws_string_new_from_cursor(secure_tunnel->allocator, &message_view->service_id);
+
+        if (secure_tunnel->config->service_id_1 != NULL &&
+            aws_string_compare(secure_tunnel->config->service_id_1, service_id) == 0) {
+            stream_id = secure_tunnel->config->service_id_1_stream_id;
+        } else if (
+            secure_tunnel->config->service_id_2 != NULL &&
+            aws_string_compare(secure_tunnel->config->service_id_2, service_id) == 0) {
+            stream_id = secure_tunnel->config->service_id_2_stream_id;
+        } else if (
+            secure_tunnel->config->service_id_3 != NULL &&
+            aws_string_compare(secure_tunnel->config->service_id_3, service_id) == 0) {
+            stream_id = secure_tunnel->config->service_id_3_stream_id;
+        } else {
+            /* service_id doesn't match any existing service id*/
+            AWS_LOGF_DEBUG(
```

Comment:
Should this be a fatal protocol error?

Reply:
Tricky. It's possible the user uses a service id that isn't one the current secure tunnel supports. The protocol/specs for secure tunnel doesn't explicitly say anything about what to do in this scenario so I'm just letting it slide with a debug message. I could maybe swap it to a Warn level message instead of Debug.